src/agent_observer.py

```diff
@@ -0,0 +1,104 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The agent relation observer module."""
+
+
+import logging
+import os
+from pathlib import Path
+
+import ops
+
+import service
+from charm_state import AGENT_RELATION, State
+
+logger = logging.getLogger()
+JENKINS_WORKDIR = Path("/var/snap/jenkins-agent/common/agent")
+AGENT_READY_PATH = Path(JENKINS_WORKDIR / ".ready")
+
+
+class Observer(ops.Object):
+    """The Jenkins agent relation observer."""
+
+    def __init__(
+        self,
+        charm: ops.CharmBase,
+        state: State,
+        jenkins_agent_service: service.JenkinsAgentService,
+    ):
+        """Initialize the observer and register event handlers.
+
+        Args:
+            charm: The parent charm to attach the observer to.
+            state: The charm state.
+            jenkins_agent_service: Service manager that controls Jenkins agent service.
+        """
+        super().__init__(charm, "agent-observer")
+        self.charm = charm
+        self.state = state
+        self.jenkins_agent_service = jenkins_agent_service
+        charm.framework.observe(
+            charm.on[AGENT_RELATION].relation_joined, self._on_agent_relation_joined
+        )
+        charm.framework.observe(
+            charm.on[AGENT_RELATION].relation_changed, self._on_agent_relation_changed
+        )
+        charm.framework.observe(
+            charm.on[AGENT_RELATION].relation_departed, self._on_agent_relation_departed
+        )
+
+    def _on_agent_relation_joined(self, event: ops.RelationJoinedEvent) -> None:
+        """Handle agent relation joined event.
+
+        Args:
+            event: The event fired when an agent has joined the relation.
+        """
+        self.charm.unit.status = ops.MaintenanceStatus(
+            f"Setting up '{event.relation.name}' relation."
+        )
+
+        relation_data = self.state.agent_meta.get_jenkins_agent_v0_interface_dict()
+        logger.debug("Agent relation data set: %s", relation_data)
+        event.relation.data[self.charm.unit].update(relation_data)
+
+    def _on_agent_relation_changed(self, event: ops.RelationChangedEvent) -> None:
+        """Handle agent relation changed event.
+
+        Args:
+            event: The event fired when the agent relation data has changed.
+        """
+        logger.info("%s relation changed.", event.relation.name)
+
+        # Check if the pebble service has started and set agent ready.
+        if os.path.exists(str(AGENT_READY_PATH)) and self.jenkins_agent_service.is_active:
+            logger.warning("Given agent already registered. Skipping.")
+            return
+
+        if not self.state.agent_relation_credentials:
```

Comment:
What would be an example of where the relation data is not complete please?

Reply:
This could happen in a scenario where `jenkins-k8s` is related to `jenkins-agent` with 2 or more units. Since jenkins puts the connection credentials into the relation databag for each unit, one at a time, `agent-relation-changed` can fire multiple times for each agent unit, without necessarily containing the credentials needed to establish a connection.